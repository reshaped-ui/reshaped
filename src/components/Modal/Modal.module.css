.root {
	transition: var(--rs-easing-accelerate) var(--rs-duration-medium);
	transition-property: transform, opacity;
	background: var(--rs-color-background-elevation-overlay);
	color: var(--rs-color-foreground-neutral);
	box-shadow: var(--rs-shadow-overlay);
	will-change: transform;

	&:focus-visible {
		outline: none;
		box-shadow: var(--rs-focus-shadow);
	}
}

.--dragging {
	transition: none;
}

@responsive .root {
	@variable --rs-modal-size calc(var(--rs-unit-x1) * 100);
}

@responsive .--position {
	@value center {
		--rs-modal-size-s: calc(var(--rs-unit-x1) * 100);

		width: var(--rs-modal-size);
		height: auto;
		max-width: calc(100vw - var(--rs-unit-x8));
		max-height: none;
		margin: var(--rs-unit-x4);
		inset: 0;
		transform: scale(0.96);
		position: relative;
		border-radius: var(--rs-unit-radius-large);
		overflow: hidden;
		opacity: 0;

		&.--active,
		[dir="rtl"] &.--active {
			transform: translate(0, 0) scale(1) !important;
			opacity: 1;
		}
	}

	@value bottom {
		--rs-modal-size-s: auto;

		width: 100%;
		height: var(--rs-modal-size);
		max-width: 100%;
		max-height: calc(100vh - var(--rs-unit-x4));
		margin: 0;
		margin-top: var(--rs-unit-x4);
		inset: 0;
		inset-block-start: initial;
		transform: translate(0, 100%);
		position: fixed;
		border-radius: var(--rs-unit-radius-large) var(--rs-unit-radius-large) 0 0;
		overflow: auto;
		opacity: 1;

		&.--active,
		[dir="rtl"] &.--active {
			transform: translate(0, max(var(--rs-modal-drag, 0px), 0px)) scale(1) !important;
		}
	}

	@value start {
		--rs-modal-size-s: calc(var(--rs-unit-x1) * 100);

		width: var(--rs-modal-size);
		height: 100%;
		max-width: calc(100vw - var(--rs-unit-x4));
		max-height: 100%;
		margin: 0;
		margin-inline-end: var(--rs-unit-x4);
		inset: 0;
		inset-inline-end: initial;
		transform: translate(-100%, 0);
		position: fixed;
		border-radius: 0;
		overflow: auto;
		opacity: 1;

		[dir="rtl"] & {
			transform: translate(100%, 0);
		}

		&.--active,
		[dir="rtl"] &.--active {
			transform: translate(min(var(--rs-modal-drag, 0px), 0px), 0) scale(1) !important;
		}
	}

	@value end {
		--rs-modal-size-s: calc(var(--rs-unit-x1) * 100);

		width: var(--rs-modal-size);
		height: 100%;
		max-width: calc(100vw - var(--rs-unit-x4));
		max-height: 100%;
		margin: 0;
		margin-inline-start: var(--rs-unit-x4);
		inset: 0;
		inset-inline-start: initial;
		transform: translate(100%, 0);
		position: fixed;
		border-radius: 0;
		overflow: auto;
		opacity: 1;

		[dir="rtl"] & {
			transform: translate(-100%, 0);
		}

		&.--active,
		[dir="rtl"] &.--active {
			transform: translate(max(var(--rs-modal-drag, 0px), 0px), 0) scale(1) !important;
		}
	}

	@value full-screen {
		--rs-modal-size-s: 100%;

		width: 100%;
		height: 100%;
		max-width: 100%;
		max-height: 100%;
		margin: 0;
		inset: 0;
		inset-block-start: initial;
		transform: translate(0, var(--rs-unit-x4));
		position: fixed;
		border-radius: 0;
		overflow: auto;
		opacity: 0;

		&.--active,
		[dir="rtl"] &.--active {
			transform: translate(0, 0) !important;
			opacity: 1;
		}
	}
}

.--active,
[dir="rtl"] .--active {
	transition-timing-function: var(--rs-easing-decelerate);
}
