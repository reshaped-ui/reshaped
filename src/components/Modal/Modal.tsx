"use client";

import React from "react";
import {
	classNames,
	responsiveVariables,
	responsiveClassNames,
	responsivePropDependency,
} from "utilities/props";
import { enableUserSelect, disableUserSelect } from "utilities/dom";
import { enableScroll, disableScroll } from "utilities/scroll";
import useResponsiveClientValue from "hooks/useResponsiveClientValue";
import Text from "components/Text";
import Overlay from "components/Overlay";
import useElementId from "hooks/useElementId";
import type * as T from "./Modal.types";
import s from "./Modal.module.css";
import getPaddingStyles from "styles/resolvers/padding";
import useHandlerRef from "hooks/useHandlerRef";

const DRAG_THRESHOLD = 32;
const DRAG_OPPOSITE_THRESHOLD = 100;
const DRAG_EDGE_BOUNDARY = 32;

const Context = React.createContext<T.Context>({
	id: "",
	titleMounted: false,
	setTitleMounted: () => {},
	subtitleMounted: false,
	setSubtitleMounted: () => {},
});
const useModal = () => React.useContext(Context);

const ModalTitle: React.FC<T.TitleProps> = (props) => {
	const { children } = props;
	const { id, setTitleMounted } = useModal();

	React.useEffect(() => {
		setTitleMounted(true);
		return () => setTitleMounted(false);
	}, [setTitleMounted]);

	return (
		<Text variant="featured-3" weight="bold" attributes={{ id: `${id}-title` }}>
			{children}
		</Text>
	);
};

const ModalSubtitle: React.FC<T.SubtitleProps> = (props) => {
	const { children } = props;
	const { id, setSubtitleMounted } = useModal();

	React.useEffect(() => {
		setSubtitleMounted(true);
		return () => setSubtitleMounted(false);
	}, [setSubtitleMounted]);

	return (
		<Text variant="body-3" color="neutral-faded" attributes={{ id: `${id}-subtitle` }}>
			{children}
		</Text>
	);
};

const Modal: React.FC<T.Props> & {
	Title: typeof ModalTitle;
	Subtitle: typeof ModalSubtitle;
} = (props) => {
	const {
		children,
		onClose,
		onOpen,
		onAfterClose,
		onAfterOpen,
		active,
		size,
		padding = 4,
		position = "center",
		overflow,
		transparentOverlay,
		blurredOverlay,
		ariaLabel,
		autoFocus = true,
		disableSwipeGesture,
		disableCloseOnOutsideClick,
		containerRef,
		overlayClassName,
		className,
		attributes,
	} = props;
	const onCloseRef = useHandlerRef(onClose);
	const id = useElementId();
	const clientPosition = useResponsiveClientValue(position)!;
	const [titleMounted, setTitleMounted] = React.useState(false);
	const [subtitleMounted, setSubtitleMounted] = React.useState(false);
	const [dragging, setDragging] = React.useState(false);
	const internalRootRef = React.useRef<HTMLDivElement>(null);
	const rootRef = attributes?.ref || internalRootRef;
	const dragStartCoordinatesRef = React.useRef({ x: 0, y: 0 });
	const dragLastCoordinateRef = React.useRef(0);
	const dragDistanceRef = React.useRef(0);
	const dragDirectionRef = React.useRef(0);
	const [dragDistance, setDragDistance] = React.useState(0);
	const [hideProgress, setHideProgress] = React.useState(0);
	const paddingStyles = getPaddingStyles(padding);

	const value = React.useMemo(
		() => ({
			titleMounted,
			setTitleMounted,
			subtitleMounted,
			setSubtitleMounted,
			id,
		}),
		[id, subtitleMounted, titleMounted]
	);

	const resetDragData = () => {
		dragStartCoordinatesRef.current = { x: 0, y: 0 };
		dragLastCoordinateRef.current = 0;
		dragDirectionRef.current = 0;
		setDragDistance(0);
	};

	const handleDragStart = (e: React.TouchEvent) => {
		if (disableSwipeGesture) return;

		// Prevent swipe to close from happening when user is working with text selection
		if (window.getSelection()?.toString()) return;

		let currentEl = e.target as HTMLElement | null;
		const rootEl = rootRef.current;

		while (currentEl && (currentEl === rootEl || rootEl?.contains(currentEl))) {
			// Prioritize scrolling over modal swiping
			if (currentEl.scrollTop !== 0 || currentEl.scrollLeft !== 0) return;
			// Start dragging only when starting on static elements
			if (currentEl.matches("input,textarea")) return;

			currentEl = currentEl ? currentEl.parentElement : null;
		}

		// Prevent the drag handling when browser tab swiping is triggering
		if (clientPosition === "start" && e.targetTouches[0].clientX < DRAG_EDGE_BOUNDARY) return;

		disableUserSelect();
		disableScroll();
		setDragging(true);
	};

	// Once modal is closed - reset its drag data
	const handleTransitionEnd = (e: React.TransitionEvent) => {
		if (active) return;
		if (e.propertyName !== "transform") return;
		if (e.currentTarget !== e.target) return;

		resetDragData();
	};

	React.useEffect(() => {
		if (!dragging) return;

		const handleDragEnd = () => {
			enableUserSelect();
			enableScroll();
			setDragging(false);

			// Close only when dragging in the closing direction
			// Changing to a different direction will keep the modal opened
			const shouldClose =
				clientPosition === "start" ? dragDirectionRef.current < 0 : dragDirectionRef.current > 0;

			if (Math.abs(dragDistanceRef.current) > DRAG_THRESHOLD && shouldClose) {
				onCloseRef.current?.({ reason: "drag" });
			} else {
				resetDragData();
			}
		};

		const handleDrag = (e: TouchEvent) => {
			if (!dragging || clientPosition === "center") return;
			if (rootRef.current?.scrollTop !== 0 || rootRef.current?.scrollLeft !== 0) return;

			const target = e.targetTouches[0];
			const coordinate = { x: target.clientX, y: target.clientY };
			const key = clientPosition === "bottom" ? "y" : "x";
			const oppositeKey = clientPosition === "bottom" ? "x" : "y";

			// Save the initial coordinates
			if (!dragStartCoordinatesRef.current[key]) {
				dragStartCoordinatesRef.current = coordinate;
				dragLastCoordinateRef.current = coordinate[key];
			}

			const next = Math.abs(coordinate[key] - dragStartCoordinatesRef.current[key]);
			const nextPerpendicular = Math.abs(
				coordinate[oppositeKey] - dragStartCoordinatesRef.current[oppositeKey]
			);

			// For start/end drawers - ignore the swiping
			// If user is scrolling vertically more than swiping
			if (
				position !== "bottom" &&
				(next < nextPerpendicular || nextPerpendicular > DRAG_OPPOSITE_THRESHOLD)
			) {
				dragLastCoordinateRef.current = coordinate[key];
				return;
			}

			dragDirectionRef.current = coordinate[key] - dragLastCoordinateRef.current;
			dragLastCoordinateRef.current = coordinate[key];

			setDragDistance((prev) =>
				clientPosition === "start"
					? Math.min(0, prev + dragDirectionRef.current)
					: Math.max(0, prev + dragDirectionRef.current)
			);
		};

		document.addEventListener("touchmove", handleDrag, { passive: true });
		document.addEventListener("touchend", handleDragEnd, { passive: true });

		return () => {
			document.removeEventListener("touchmove", handleDrag);
			document.removeEventListener("touchend", handleDragEnd);
		};
	}, [dragging, clientPosition, onCloseRef, position, rootRef]);

	// Syncing distance to the ref to avoid having a dependency on dragDistance in handleDragEnd
	React.useEffect(() => {
		const rootEl = rootRef.current;

		if (!rootEl || !clientPosition) return;

		const isInline = ["start", "end"].includes(clientPosition);

		const size = isInline ? rootEl.clientWidth : rootEl.clientHeight;
		const progress = Math.abs(dragDistance) / size;

		setHideProgress(progress / 2);
		dragDistanceRef.current = dragDistance;
	}, [dragDistance, clientPosition, rootRef]);

	return (
		<Overlay
			onClose={onClose}
			onOpen={onOpen}
			onAfterClose={onAfterClose}
			onAfterOpen={onAfterOpen}
			disableCloseOnClick={disableCloseOnOutsideClick}
			active={active}
			transparent={transparentOverlay || hideProgress}
			blurred={blurredOverlay}
			overflow={responsivePropDependency(position, (p) => (p === "center" ? "auto" : "hidden"))}
			className={overlayClassName}
			containerRef={containerRef}
			attributes={{
				onTouchStart: handleDragStart,
			}}
		>
			{({ active }) => {
				const rootClassNames = classNames(
					s.root,
					className,
					active && s["--active"],
					dragging && s["--dragging"],
					overflow && s[`--overflow-${overflow}`],
					containerRef && s["--contained"],
					responsiveClassNames(s, "--position", position)
				);

				return (
					<Context.Provider value={value}>
						{}
						<div
							{...attributes}
							style={
								{
									...paddingStyles?.variables,
									...responsiveVariables("--rs-modal-size", size),
									"--rs-modal-drag":
										Math.abs(dragDistance) < DRAG_THRESHOLD
											? "0px"
											: `${
													dragDistance + DRAG_THRESHOLD * (clientPosition === "start" ? 1 : -1)
												}px`,
								} as React.CSSProperties
							}
							aria-labelledby={titleMounted ? `${id}-title` : undefined}
							aria-describedby={subtitleMounted ? `${id}-subtitle` : undefined}
							aria-label={ariaLabel || attributes?.["aria-label"]}
							className={rootClassNames}
							aria-modal="true"
							role="dialog"
							tabIndex={!autoFocus ? -1 : undefined}
							ref={rootRef}
							onTransitionEnd={handleTransitionEnd}
						>
							{children}
						</div>
					</Context.Provider>
				);
			}}
		</Overlay>
	);
};

Modal.Title = ModalTitle;
Modal.Subtitle = ModalSubtitle;

Modal.displayName = "Modal";
ModalTitle.displayName = "Modal.Title";
ModalSubtitle.displayName = "Modal.Subtitle";

export default Modal;
