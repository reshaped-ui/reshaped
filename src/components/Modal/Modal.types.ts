import type React from "react";
import type * as G from "types/global";
import type { OverlayProps } from "components/Overlay";

export type Context = {
	id: string;
	titleMounted: boolean;
	setTitleMounted: (b: boolean) => void;
	subtitleMounted: boolean;
	setSubtitleMounted: (b: boolean) => void;
};

export type TitleProps = {
	children: React.ReactNode;
};

export type SubtitleProps = {
	children: React.ReactNode;
};

export type Props = {
	children?: React.ReactNode;
	position?: G.Responsive<"center" | "end" | "bottom" | "start" | "full-screen">;
	size?: G.Responsive<string>;
	padding?: G.Responsive<number>;
	overflow?: "visible";
	transparentOverlay?: boolean;
	blurredOverlay?: boolean;
	disableSwipeGesture?: boolean;
	disableCloseOnOutsideClick?: boolean;
	autoFocus?: boolean;
	ariaLabel?: string;
	className?: G.ClassName;
	overlayClassName?: G.ClassName;
	attributes?: G.Attributes<"div"> & { ref?: React.RefObject<HTMLDivElement> };
} & Pick<OverlayProps, "onClose" | "onOpen" | "active">;
