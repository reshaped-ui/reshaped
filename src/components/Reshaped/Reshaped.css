:root {
	--rs-z-index-raised: 5;
	--rs-z-index-flyout: 80;
	--rs-z-index-fixed: 90;
	--rs-z-index-overlay: 100;
	--rs-z-index-notification: 110;
}

body,
h1,
h2,
h3,
h4,
h5,
h6,
p,
ul,
ol,
li,
figure,
figcaption,
blockquote,
dl,
dd {
	margin: 0;
	padding: 0;
}

ul[class],
ol[class] {
	list-style: none;
}

textarea {
	resize: vertical;
}

table {
	border-collapse: collapse;
	border-spacing: 0;
}

fieldset {
	border: 0;
	margin: 0;
	padding: 0;
}

img {
	max-width: 100%;
	display: block;
}

input,
button,
textarea,
select {
	font: inherit;
}

/* 
	Setting background for Windows browsers that have white background by default 
	which doesn't work well in dark mode
	*/
option {
	background: var(--rs-color-background-elevation-base);
}

label {
	cursor: pointer;
}

textarea::placeholder,
input::placeholder {
	opacity: 0.5;
	color: var(--rs-color-foreground-neutral-faded);
}

html {
	/* https://developer.mozilla.org/en-US/docs/Web/CSS/text-size-adjust */
	-webkit-text-size-adjust: 100%;
	-webkit-font-smoothing: antialiased;

	/* https://developer.mozilla.org/en-US/docs/Web/CSS/text-rendering */
	text-rendering: optimizelegibility;

	/* https://developer.mozilla.org/en-US/docs/Web/CSS/touch-action#values */
	touch-action: manipulation;
}

* {
	box-sizing: border-box;
}

html,
body {
	color: var(--rs-color-foreground-neutral);
	background: var(--rs-color-background-page);
	height: 100%;
	scroll-behavior: smooth;
}

body,
[data-rs-theme] {
	font-family: var(--rs-font-family-body);
	font-weight: var(--rs-font-weight-regular);
}

[data-rs-color-mode="light"] {
	color-scheme: light;
}

[data-rs-color-mode="dark"] {
	color-scheme: dark;
}

@media (prefers-reduced-motion: reduce) {
	* {
		animation-duration: 0.01ms !important;
		animation-iteration-count: 1 !important;
		transition-duration: 0.01ms !important;
		scroll-behavior: auto !important;
	}
}

[data-rs-no-transition] *,
[data-rs-no-transition] *::before,
[data-rs-no-transition] *::after {
	transition: none !important;
}

/* Used for resetting headless component styles */
@layer rs.reset;
