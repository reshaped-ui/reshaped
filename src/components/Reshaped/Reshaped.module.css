.root {
	display: contents;
	color: inherit;
}
