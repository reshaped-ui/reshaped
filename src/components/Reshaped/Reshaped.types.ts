import type React from "react";
import type { GlobalColorModeProps, ThemeProps } from "components/Theme";
import type { ToastProviderProps } from "components/Toast";
import type * as G from "types/global";

export type Props = {
	/** Node for inserting children */
	children?: React.ReactNode;
	/** Theme of the application, enables controlled mode */
	theme?: NonNullable<ThemeProps["name"]>;
	/** Default theme of the application, enables uncontrolled mode */
	defaultTheme?: NonNullable<ThemeProps["defaultName"]>;
	/** Default content direction of the application */
	defaultRTL?: boolean;
	/** Color mode of the application, enables controlled mode */
	colorMode?: GlobalColorModeProps["mode"];
	/** Default color mode of the application, enables uncontrolled mode */
	defaultColorMode?: GlobalColorModeProps["defaultMode"];
	/** Default viewport size of the application */
	defaultViewport?: G.Viewport;
	/** Global options for the ToastProvider */
	toastOptions?: ToastProviderProps["options"];
	/** Enable scoped mode for applications not using Reshaped provider at the application root */
	scoped?: boolean;
	/** Additional classname for the root element */
	className?: G.ClassName;
	/** Additional attributes for the root element */
	attributes?: G.Attributes<"div">;
};
