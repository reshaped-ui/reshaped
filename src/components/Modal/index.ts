import Modal, { ModalTitle, ModalSubtitle } from "./Modal";
import type * as T from "./Modal.types";

const ModalRoot = Modal as React.FC<T.Props> & {
	Title: typeof ModalTitle;
	Subtitle: typeof ModalSubtitle;
};

ModalRoot.Title = ModalTitle;
ModalRoot.Subtitle = ModalSubtitle;

export default ModalRoot;
export type { Props as ModalProps } from "./Modal.types";
