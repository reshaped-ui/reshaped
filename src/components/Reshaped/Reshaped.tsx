"use client";

import React from "react";
import { classNames } from "utilities/helpers";
import { GlobalColorMode, PrivateTheme } from "components/Theme";
import { ToastProvider } from "components/Toast";
import { useGlobalColorMode } from "components/Theme/useTheme";
import {
	SingletonEnvironmentContext,
	useSingletonEnvironment,
} from "hooks/_private/useSingletonEnvironment";
import { SingletonKeyboardModeProvider } from "hooks/_private/useSingletonKeyboardMode";
import { SingletonHotkeysProvider } from "hooks/_private/useSingletonHotkeys";
import type * as T from "./Reshaped.types";
import "./Reshaped.css";
import s from "./Reshaped.module.css";

const ReshapedInner = (props: T.Props) => {
	const { children, defaultRTL, defaultViewport = "s", toastOptions } = props;
	const rtlState = useSingletonEnvironment(defaultRTL);

	return (
		<SingletonKeyboardModeProvider>
			<SingletonEnvironmentContext.Provider value={{ rtl: rtlState, defaultViewport }}>
				<SingletonHotkeysProvider>
					<ToastProvider options={toastOptions}>{children}</ToastProvider>
				</SingletonHotkeysProvider>
			</SingletonEnvironmentContext.Provider>
		</SingletonKeyboardModeProvider>
	);
};

const Reshaped = (props: T.Props) => {
	const { theme, defaultTheme = "reshaped", defaultColorMode, scoped, className } = props;
	const rootClassNames = classNames(s.root, className);
	const scopeRef = React.useRef<HTMLDivElement>(null);
	const parentGlobalColorMode = useGlobalColorMode();

	return (
		<GlobalColorMode
			defaultMode={defaultColorMode || parentGlobalColorMode.mode || "light"}
			scopeRef={!!parentGlobalColorMode && scoped ? scopeRef : undefined}
		>
			<PrivateTheme
				name={theme}
				defaultName={defaultTheme}
				className={rootClassNames}
				scoped={scoped}
				scopeRef={!!parentGlobalColorMode && scoped ? scopeRef : undefined}
			>
				<ReshapedInner {...props}>{props.children}</ReshapedInner>
			</PrivateTheme>
		</GlobalColorMode>
	);
};

export default Reshaped;
