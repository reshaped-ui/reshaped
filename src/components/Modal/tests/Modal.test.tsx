import React from "react";
import { render, screen, waitFor, fireEvent } from "@testing-library/react";
import { userEvent } from "@testing-library/user-event";
import Modal from "components/Modal";
import Reshaped from "components/Reshaped";

const fixtures = {
	content: "Content",
	title: "Title",
	subtitle: "Subtitle",
	className: "test-className",
	overlayClassName: "test-overlay-className",
	id: "test-id",
};

Object.defineProperty(window, "matchMedia", {
	writable: true,
	value: jest.fn().mockImplementation((query) => ({
		matches: false,
		media: query,
		onchange: null,
		addEventListener: jest.fn(),
		removeEventListener: jest.fn(),
		dispatchEvent: jest.fn(),
	})),
});

describe("Components/Modal", () => {
	test("doesn't render children", () => {
		render(
			<Reshaped>
				<Modal>{fixtures.content}</Modal>
			</Reshaped>
		);

		expect(screen.queryByText(fixtures.content)).not.toBeInTheDocument();
	});

	test("renders correctly when active", () => {
		render(
			<Reshaped>
				<Modal active>
					<Modal.Title>{fixtures.title}</Modal.Title>
					<Modal.Subtitle>{fixtures.subtitle}</Modal.Subtitle>
					{fixtures.content}
				</Modal>
			</Reshaped>
		);

		const elModal = screen.getByRole("dialog");
		const elTitle = screen.getByText(fixtures.title);
		const elSubtitle = screen.getByText(fixtures.subtitle);
		const titleId = elTitle.getAttribute("id");
		const subtitleId = elSubtitle.getAttribute("id");

		expect(elModal).toBeInTheDocument();
		expect(elTitle).toBeInTheDocument();
		expect(elSubtitle).toBeInTheDocument();
		expect(titleId).toMatchSnapshot();
		expect(subtitleId).toMatchSnapshot();

		expect(elModal).toHaveAttribute("aria-modal", "true");
		expect(elModal).toHaveAttribute("aria-labelledby", titleId);
		expect(elModal).toHaveAttribute("aria-describedby", subtitleId);
	});

	test("works as controlled", async () => {
		const handleCloseMock = jest.fn();

		const Component = () => {
			const [active, setActive] = React.useState(true);

			const handleOpen = () => {
				setActive(true);
			};

			const handleClose = () => {
				setActive(false);
				handleCloseMock();
			};

			return (
				<Reshaped>
					<button type="button" onClick={handleOpen}>
						Open
					</button>
					<Modal active={active} onClose={handleClose}>
						{fixtures.content}
					</Modal>
				</Reshaped>
			);
		};
		render(<Component />);

		expect(screen.queryByText(fixtures.content)).toBeInTheDocument();

		await userEvent.keyboard("{Escape}");

		/**
		 * Need to trigger transition event on the correct element
		 * since Overlay ignores event bubbling to prevent it from closing based on other elements transitionend
		 */
		const overlayEl = screen.getAllByRole("button")[1];
		fireEvent.transitionEnd(overlayEl, {
			propertyName: "transform",
			pseudoElement: "::after",
		});

		await waitFor(() => {
			expect(handleCloseMock).toHaveBeenCalledTimes(1);

			// TODO: Flaky behavior, make it stable
			// expect(screen.queryByText(fixtures.content)).not.toBeInTheDocument();
		});

		const elButton = screen.getByText("Open");

		await userEvent.click(elButton);

		expect(screen.getByText(fixtures.content)).toBeInTheDocument();
	});

	test("works with className and attributes", () => {
		render(
			<Reshaped>
				<Modal
					active
					className={fixtures.className}
					attributes={{ id: fixtures.id }}
					overlayClassName={fixtures.overlayClassName}
				>
					{fixtures.content}
				</Modal>
			</Reshaped>
		);

		const elModal = screen.getByRole("dialog");
		const elOverlay = screen.getByRole("button");

		expect(elModal).toHaveClass(fixtures.className);
		expect(elModal).toHaveAttribute("id", fixtures.id);
		expect(elOverlay).toHaveClass(fixtures.overlayClassName);
	});
});
