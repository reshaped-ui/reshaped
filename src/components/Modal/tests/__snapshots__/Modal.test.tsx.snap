// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Components/Modal renders correctly when active 1`] = `":r2:-title"`;

exports[`Components/Modal renders correctly when active 2`] = `":r2:-subtitle"`;
